### PR TITLE
Addons bookmark refresh fix

### PR DESCRIPTION
**Pull request by eht16**
Update bookmark list also on any line changes in the document

In addition to updating the bookmark list on document notebook tab change
and adding/removing line markers, also rebuild the bookmark list
when the text in the document changes with lines added or deleted.

Should fix SF bugs 129 and 39.

This implements some ideas mentioned in the above bug reports like updating the bookmark list
in an idle callback. Additionally, the list store is detach from the GtkTreeView to increase performance on big bookmark lists.

Maybe it is still worth to add a refresh button (or rather a context menu item to refresh) but I would try it without at first.

There is one issue left: if the `Transpose current line` feature is used, the bookmarks don't update. But this is not a problem in the plugin but rather in either Geany or Scintilla as the marker also stays at the "old" line. This should be managed separately.

### CODE REVIEW

**Comment by codebrainz** [comment]
LGTM, though I didn't test.

File: addons/src/ao_bookmarklist.c [hunk 1]
```diff
@@ -62,8 +62,16 @@ struct _AoBookmarkListPrivate
...
 	gint		 search_line;
 	GtkTreeIter	*search_iter;
+
+	guint refresh_idle_source_id;
 };
 
+typedef struct
+{
+	AoBookmarkList *bm;
+	GeanyDocument *document;
```
Comment (b4n) [inline addons/src/ao_bookmarklist.c]: you might want to use Document IDs as you delay the reference to the document.  It probably doesn't matter much in practice though, it's highly unlikely a document is reused during the delay.

**Comment by b4n** [comment]
Otherwise, as @codebrainz, I didn't test but it LGTM

**Comment by eht16** [comment]
Thanks for the feedback.

I changed the document referencing to IDs as suggested.

Refreshing the treeview with about 5000 markers set takes about 40-50ms with de- and reattching the model. Without detaching, refreshing takes about 60-70ms, on my machine.
So it is actually not that big difference and I doubt users will click that many markers. I cheated a bit by programatically setting them on startup :).

So, as the difference is actually not really big, I removed commit 0d130be42b3e7968c597c85a1782b1e34d69d5a8. However, I could not notice any flickering, with or without the commit.

File: addons/src/ao_bookmarklist.c [hunk 2]
```diff
@@ -401,14 +409,17 @@ void ao_bookmark_list_activate(AoBookmarkList *bm)
...
+static gboolean update_bookmark_list_delayed(gpointer data)
 {
 	gint line_nr = 0;
 	gint mask = 1 << 1;
-	ScintillaObject *sci = doc->editor->sci;
+	AoBookmarkListRefreshContainer *container = data;
+	AoBookmarkList *bm = container->bm;
 	AoBookmarkListPrivate *priv = AO_BOOKMARK_LIST_GET_PRIVATE(bm);
+	GeanyDocument *doc = document_find_by_id(container->document_id);
+	ScintillaObject *sci = doc->editor->sci;
```
Comment (b4n) [inline addons/src/ao_bookmarklist.c]: logic error: you dereference `doc` here, but check whether it's NULL on the next line.

Comment (db0451) [inline addons/src/ao_bookmarklist.c]: I assume the fix is this?

GeanyDocument *doc = document_find_by_id(container->document_id);

	if (doc != NULL && priv->enable_bookmarklist)
	{
		ScintillaObject *sci = doc->editor->sci;
// etc.

Comment (b4n) [inline addons/src/ao_bookmarklist.c]: yes

Comment (eht16) [inline addons/src/ao_bookmarklist.c]: Oops. Fixed in 8ab3ee28baca96e683b92a585b9ab226196b170f.

File: addons/src/ao_bookmarklist.c [hunk 3]
```diff
@@ -401,15 +409,18 @@ void ao_bookmark_list_activate(AoBookmarkList *bm)
...
-	ScintillaObject *sci = doc->editor->sci;
+	AoBookmarkListRefreshContainer *container = data;
+	AoBookmarkList *bm = container->bm;
 	AoBookmarkListPrivate *priv = AO_BOOKMARK_LIST_GET_PRIVATE(bm);
+	GeanyDocument *doc = document_find_by_id(container->document_id);
 
-	if (priv->enable_bookmarklist)
+	if (priv->enable_bookmarklist && doc != NULL)
 	{
+		ScintillaObject *sci = doc->editor->sci;
```
Comment (b4n) [inline addons/src/ao_bookmarklist.c]: style probably asks for a blank line here (really no biggie, and your call actually)

File: addons/src/ao_bookmarklist.c [hunk 4]
```diff
@@ -417,6 +428,26 @@ void ao_bookmark_list_update(AoBookmarkList *bm, GeanyDocument *doc)
...
+
+	g_free(container);
+	priv->refresh_idle_source_id = 0;
+	return FALSE;
+}
+
+
+void ao_bookmark_list_update(AoBookmarkList *bm, GeanyDocument *doc)
+{
+	AoBookmarkListPrivate *priv = AO_BOOKMARK_LIST_GET_PRIVATE(bm);
```
Comment (b4n) [inline addons/src/ao_bookmarklist.c]: here too

File: addons/src/ao_bookmarklist.c [hunk 5]
```diff
@@ -417,6 +428,26 @@ void ao_bookmark_list_update(AoBookmarkList *bm, GeanyDocument *doc)
...
+	if (priv->refresh_idle_source_id == 0)
+	{
+		AoBookmarkListRefreshContainer *container = g_new0(AoBookmarkListRefreshContainer, 1);
```
Comment (b4n) [inline addons/src/ao_bookmarklist.c]: and here

**Comment by b4n** [comment]
LGTM

**Comment by eht16** [comment]
I added https://github.com/eht16/geany-plugins/commit/e8c0dccee3a0a1688325939f1640c455498e03b0 to add the newlines and squashed unimportant commits together.

**Comment by eht16** [comment]
And it seems I pulled too many commits into the rebase. Can this get problematic when merging?

**Comment by b4n** [comment]
yes, it very well might.  But you should be able to fix that fairly easily to fix: check master out, pull (fast-forward, `pull --ff-only`) from upstream, and rebase your branch on top of that.  Or you can also just `rebase -i 8234f7c^` and remove the two offending/duplicate commits, only keeping yours.  As it's a separate branch, there won't be no problem.

**Comment by b4n** [comment]
@eht16 if you're interested, I just did the second solution as you didn't seem to update this already: https://github.com/b4n/geany-plugins/commits/eht16/addons_bookmark_refresh_fix ;)

**Comment by eht16** [comment]
Thanks again for your help and sorry for unnecessary mistakes in code and GIT usage :(.

**Comment by b4n** [comment]
np, that's what annoying reviews are for :)

**Pull request merged**